README.md
Django Opensearch DSL
=====================

[![PyPI Version](https://badge.fury.io/py/django-opensearch-dsl.svg)](https://badge.fury.io/py/django-opensearch-dsl)
[![Documentation Status](https://readthedocs.org/projects/django-opensearch-dsl/badge/?version=latest)](https://django-opensearch-dsl.readthedocs.io/en/latest/?badge=latest)
![Tests](https://github.com/Codoc-os/django-opensearch-dsl/workflows/Tests/badge.svg)
[![Python 3.7+](https://img.shields.io/badge/Python-3.7+-brightgreen.svg)](#)
[![Django 2.2+, 3.2+, 4.0+](https://img.shields.io/badge/Django-2.2+,%203.2+,%204.0+-brightgreen.svg)](#)
[![License Apache 2](https://img.shields.io/badge/license-Apache%202-brightgreen.svg)](https://github.com/Codoc-os/django-opensearch-dsl/blob/master/LICENSE)
[![codecov](https://codecov.io/gh/Codoc-os/django-opensearch-dsl/branch/master/graph/badge.svg)](https://codecov.io/gh/Codoc-os/django-opensearch-dsl)
[![CodeFactor](https://www.codefactor.io/repository/github/Codoc-os/django-opensearch-dsl/badge)](https://www.codefactor.io/repository/github/Codoc-os/django-opensearch-dsl)

**Django Opensearch DSL** is a package that allows the indexing of Django models in opensearch. It is built as a thin
wrapper around [`opensearch-dsl-py`](https://github.com/opensearch-project/opensearch-dsl-py)
so you can use all the features developed by the `opensearch-dsl` team.

You can view the full documentation
at [https://django-opensearch-dsl.readthedocs.io](https://django-opensearch-dsl.readthedocs.io/en/latest/).

## Features

- Based on [`opensearch-dsl-py`](https://github.com/opensearch-project/opensearch-dsl-py) so you can make queries with
  the [`Search`](https://elasticsearch-dsl.readthedocs.io/en/latest/search_dsl.html#the-search-object)
  object.
- Management commands for creating, deleting, and populating indices and documents.
- Opensearch auto mapping from Django models fields.
- Complex field type support (`ObjectField`, `NestedField`).
- Index fast using `parallel` indexing.

## Requirements

* `Python>=3.7`
* `django>=2.2`
* `opensearch-dsl~=1.0.0`

## Installation and Configuration

The easiest way to install `django-opensearch-dsl` is through `pip`:

* `pip install django-opensearch-dsl`

Then add `django_opensearch_dsl` to your `INSTALLED_APPS` settings.

You must then define `OPENSEARCH_DSL` in your Django settings.

For example:

```python
OPENSEARCH_DSL = {
    'default': {
        'hosts': 'localhost:9200'
    },
    'secure': {
        'hosts': [{"scheme": "https", "host": "192.30.255.112", "port": 9201}],
        'http_auth': ("admin", "password"),
        'timeout': 120,
    },
}
```

`OPENSEARCH_DSL` is then passed
to [`opensearch_dsl_py.connections.configure`](http://elasticsearch-dsl.readthedocs.io/en/stable/configuration.html#multiple-clusters)
.

## Create Document Classes

To index instances of the following model :

```python
# models.py

class Car(models.Model):
    name = models.CharField()
    color = models.CharField()
    description = models.TextField()
    type = models.IntegerField(choices=[
        (1, "Sedan"),
        (2, "Truck"),
        (4, "SUV"),
    ])
```

First create a subclass of [`django_opensearch_dsl.Document`](/document/) containing the subclasses `Index`
(which define the index' settings) and `Django` (which contains settings related to your django `Model`). Finally,
register the class using `registry.register_document()` decorator.

It is required to define `Document` classes inside a file named `documents.py` in your apps' directory.

```python
# documents.py

from django_opensearch_dsl import Document
from django_opensearch_dsl.registries import registry
from .models import Car


@registry.register_document
class CarDocument(Document):
    class Index:
        name = 'cars'  # Name of the Opensearch index
        settings = {  # See Opensearch Indices API reference for available settings
            'number_of_shards': 1,
            'number_of_replicas': 0
        }
        # Configure how the index should be refreshed after an update.
        # See Opensearch documentation for supported options.
        # This per-Document setting overrides settings.OPENSEARCH_DSL_AUTO_REFRESH.
        auto_refresh = False

    class Django:
        model = Car  # The model associated with this Document        
        fields = [  # The fields of the model you want to be indexed in Opensearch
            'name',
            'color',
            'description',
            'type',
        ]
        # Paginate the django queryset used to populate the index with the specified size
        # This per-Document setting overrides settings.OPENSEARCH_DSL_QUERYSET_PAGINATION.
        queryset_pagination = 5000
```


docs/CHANGELOG.md
# Changelog

## X.Y.Z (not released yet)

* New feature: zero-downtime document migration

### 0.4.1 (2022-08-16)

* `Document.update()` method now take an optional `using` argument allowing to specify an alternate
  OpenSearch connection defined in `OPENSEARCH_DSL`.
* Fix related document automatic indexation and deletion
  (Contributed by [Colin Seifer](https://github.com/Colin-Seifer)).
* Add `pre-delete` back into `BaseSignalProcessor.handle_m2m_changed()` to  properly update the
  index on M2M interactions (Contributed by [Colin Seifer](https://github.com/Colin-Seifer)).

## 0.4.0 (2022-08-04)

* Add support for related models. See [Document Classes](document.md#django-subclass) and
  [Document Field Reference](fields.md#using-prepare_field_with_related) for more information
  (Contributed by [Colin Seifer](https://github.com/Colin-Seifer)).
* `django-opensearch-dsl` now only tests supported version of Python and Django (mainstream and LTS). 
  This choice is made to:
    * Speed up development.
    * Speed up tests.
    * Reduce actions on github.
    * Encourage people to update their stack to supported (thus safer) versions.
* Drop support for Python 3.6.
* Drop support for Django 2.1, 2.2, 3.1.
* Now supports Django 4.1.
* Now supports `opensearch-dsl>=2.0 <3.00`.

## 0.3.0 (2022-06-22)

* Fixes internal links in documentation.
* Remove the need to declare a `TESTING` boolean in `settings.py`.

## 0.2.0 (2022-01-13)

* Restore auto-sync feature (still undocumented - Contributed by [David Guillot](https://github.com/David-Guillot))
* Add support to Django 4.0 (Contributed by [David Guillot](https://github.com/David-Guillot))
* Remove some python2 leftovers (Contributed by [David Guillot](https://github.com/David-Guillot))

## 0.1.2 (2021-12-14)

* Fixed 'Search.validate()'

## 0.1.0 (2021-12-11)

* Migrated to **Opensearch**
* Drop some feature such as auto-syncing signals and related models
* Replace `search_index` management command with `opensearch`.

## 0.1.0 (2021-12-11)

* Migrated to **Opensearch**
* Drop some feature such as auto-syncing signals and related models
* Replace `search_index` management command with `opensearch`.

## Before fork from `django-elasticsearch-dsl`

### 7.1.4 (2020-07-05)

* Configure Elasticsearch _id dynamically from document (#272)
* Use chain.from_iterable in for performance improvement (#278)
* Handle case where SimpleLazyObject being treated as an Iterable (#255)
* Camelcase default value in management command (#254)
* Various updates and fixup in docs (#250, #276)
* Start testing against Python 3.8 (#266)

### 7.1.1 (2019-12-26)

* Adding detailed documentation and published to Read The Docs #222
* Resolve name resolution while delete, create index (#228)
* Added support for Django 3.0. (#230)
* Removing old Elasticsearc compatibility (#219)
* Drop StringField in favor of TextField.

### 7.1.0 (2019-10-29)

* Support for Django `DecimalField` #141
* Indexing speedup by using `parallel` indexing. #213. Now you can pass `--parallel` or set `ELASTICSEARCH_DSL_PARALLEL`
  in your settings to get indexing speed boost while indexing through management command.
* Fixing name resolution in management command #206
* Small documentation fixes. #196

### 7.0.0 (2019-08-11)

* Support Elasticsearch 7.0 (See PR #176)
* Added order by to paginate queryset properly (See PR #153)
* Remove `standard` token filter from `README.md` and test files
* Various documentation fixes

### 6.4.2 (2019-07-26)

* Fix document importing path
* Update readme

### 6.4.1 (2019-06-14)

* The `DocType` import has changed to `Document`

### 6.4.0 (2019-06-01)

* Support elasticsearch-dsl>6.3.0
* Class `Meta` has changed to class `Django` (See PR #136)
* Add `register_document` decorator to register a document (See PR #136)
* Additional Bug fixing and others

### 0.5.1 (2018-11-07)

* Limit elastsearch-dsl to supported versions

### 0.5.0 (2018-04-22)

* Add Support for Elasticsearch 6 thanks to HansAdema

### Breaking Change:

* Django string fields now point to ES text field by default.
* Nothing should change for ES 2.X but if you are using ES 5.X, you may need to rebuild and/or update some of your
  documents.

### 0.4.5 (2018-04-22)

* Fix prepare with related models when deleted (See PR #99)
* Fix unwanted calls to get_instances_from_related
* Fix for empty ArrayField (CBinyenya)
* Fix nested OneToOneField when related object doesn't exist (CBinyenya)
* Update elasticsearch-dsl minimal version

### 0.4.4 (2017-12-13)

* Fix to_queryset with es 5.0/5.1

### 0.4.3 (2017-12-12)

* Fix syncing of related objects when deleted
* Add django 2.0 support

### 0.4.2 (2017-11-27)

* Convert lazy string to string before serialization
* Readme update (arielpontes)

### 0.4.1 (2017-10-17)

* Update example app with get_instances_from_related
* Typo/grammar fixes

### 0.4.0 (2017-10-07)

* Add a method on the Search class to return a django queryset from an es result
* Add a queryset_pagination option to DocType.Meta for allow the pagination of big django querysets during the index
  populating
* Remove the call to iterator method for the django queryset
* Fix DocType inheritance. The DocType is store in the registry as a class and not anymore as an instance

### 0.3.0 (2017-10-01)

* Add support for resynching ES documents if related models are updated (HansAdema)
* Better management for django FileField and ImageField
* Fix some errors in the doc (barseghyanartur, diwu1989)

### 0.2.0 (2017-07-02)

* Replace simple model signals with easier to customise signal processors (barseghyanartur)
* Add options to disable automatic index refreshes (HansAdema)
* Support defining DocType indexes through Meta class (HansAdema)
* Add option to set default Index settings through Django config (HansAdema)

### 0.1.0 (2017-05-26)

* First release on PyPI.


django_opensearch_dsl/indices.py
from copy import deepcopy
from datetime import datetime

from opensearch_dsl import Index as DSLIndex

from .apps import DODConfig
from .registries import registry


class Index(DSLIndex):
    """A specialized DSL index that can have versions."""

    VERSION_NAME_SEPARATOR = "--"

    def __init__(self, *args, **kwargs):
        """Creates an index and makes a deep copy of default index settings."""
        super(Index, self).__init__(*args, **kwargs)
        default_index_settings = deepcopy(DODConfig.default_index_settings())
        self.settings(**default_index_settings)

    def document(self, document):
        """Extend to register the document in the global document registry."""
        document = super(Index, self).document(document)
        registry.register_document(document)
        return document

    doc_type = document

    def get_versions(self, using=None):
        """Return a name-sorted list of Index objects named after this one."""
        return [
            Index(name)
            for name in sorted(
                self._get_connection(using=using)
                .indices.get(f"{self._name}{self.__class__.VERSION_NAME_SEPARATOR}*")
                .keys()
            )
        ]

    def get_version(self, version_name, using=None):
        """Return an Index that is a version of the current one."""
        for version in self.get_versions():
            if version._name == version_name:
                return version
        raise ValueError(f"Index version '{version_name}' not found for Index '{self._name}'")

    def get_active_version(self, using=None):
        """Return the concrete index that's active for this one."""
        for version in self.get_versions(using=using):
            if version.exists_alias(name=self._name):
                return version

        if self.exists():
            return self

    def create_new_version(self, using=None):
        """Return a new Index, cloned from this one, with versioned name."""
        timestamp = datetime.now().strftime("%Y%m%d%H%M%S%f")
        new_index = self.clone(f"{self._name}{self.__class__.VERSION_NAME_SEPARATOR}{timestamp}")
        new_index.create(using=using)
        return new_index

    def activate_version(self, version_name, using=None):
        """Sets an alias to the Index on the given Index version."""
        index = self.get_version(version_name)

        actions_on_aliases = [
            {"add": {"index": index._name, "alias": self._name}},
        ]

        active_version = self.get_active_version(using=using)
        if active_version:
            actions_on_aliases.insert(
                0,
                {"remove": {"index": active_version._name, "alias": self._name}},
            )

        self._get_connection(using=using).indices.update_aliases(body={"actions": actions_on_aliases})

    def __str__(self):
        return self._name


django_opensearch_dsl/management/types.py
import datetime
from typing import Union, List

from django.conf import settings
from django.utils.timezone import make_aware, is_aware

Nothing = type(Ellipsis)
Values = Union[None, int, float, datetime.datetime, str, List['Values']]


def datetime_parser(value: str) -> Union[Nothing, datetime.datetime]:
    """Try to parse the given value as a ISO 8601's datetime."""
    try:
        date = datetime.datetime.fromisoformat(value)
        if not is_aware(date):
            date = make_aware(date)
        return date
    except ValueError as e:
        return ...


def int_parser(value: str) -> Union[Nothing, int]:
    """Try to parse the given value as an integer."""
    try:
        return int(value)
    except ValueError:
        return ...


def float_parser(value: str) -> Union[Nothing, float]:
    """Try to parse the given value as a float."""
    try:
        return float(value)
    except ValueError:
        return ...


def none_parser(value: str) -> Union[Nothing, None]:
    """Try to parse the given value as a float."""
    return None if value == "" else ...


def list_parser(value: str) -> Values:
    """Try to parse the given value as a ISO 8601's datetime."""
    if "," in value:
        return [parse(v.strip()) for v in value.split(",")]
    return ...


def parse(value: str) -> Values:
    """Try to coerce a string into a different type.

    The order in which the parsers are called matters as some type might include
    other, e.g. `float_parser()` would match an `int` as a `float` if called
    before `int_parser()`.

    If no parser was able to parse the value, it is returned as a string."""
    parsers = getattr(
        settings, 'OPENSEARCH_DSL_VALUE_PARSERS',
        [none_parser, int_parser, float_parser, datetime_parser, list_parser]
    )
    for parser in parsers:
        v = parser(value)
        if v != ...:  # noqa
            return v
    return value


django_opensearch_dsl/management/commands/opensearch.py
import argparse
import functools
import operator
import os
import sys
from argparse import ArgumentParser
from collections import defaultdict
from typing import Any, Callable

import opensearchpy
from django.conf import settings
from django.core.exceptions import FieldError
from django.core.management import BaseCommand
from django.core.management.base import OutputWrapper
from django.db.models import Q
from opensearch_dsl import Index

from django_opensearch_dsl.registries import registry
from ..enums import OpensearchAction
from ..types import parse


class Command(BaseCommand):
    """Manage indices and documents."""

    help = (
        "Allow to create and delete indices, as well as indexing, updating or deleting specific "
        "documents from specific indices.\n"
    )

    def __init__(self, *args, **kwargs):  # noqa
        super(Command, self).__init__()
        self.usage = None

    def db_filter(self, parser: ArgumentParser) -> Callable[[str], Any]:
        """Return a function to parse the filters."""

        def wrap(value):  # pragma: no cover
            try:
                lookup, v = value.split("=")
                v = parse(v)
            except ValueError:
                sys.stderr.write(parser._subparsers._group_actions[0].choices["document"].format_usage())  # noqa
                sys.stderr.write(
                    f"manage.py index: error: invalid filter: '{value}' (filter must be formatted as "
                    f"'[Field Lookups]=[value]')\n",
                )
                exit(1)
            return lookup, v  # noqa

        return wrap

    def __list_index(self, **options):  # noqa pragma: no cover
        """List all known index and indicate whether they are created or not."""
        indices = registry.get_indices()
        result = defaultdict(list)
        for index in indices:
            module = index._doc_types[0].__module__.split(".")[-2]  # noqa
            exists = index.exists()
            checkbox = f"[{'X' if exists else ' '}]"
            versions = index.get_versions()
            if versions:
                details = ". Following index versions exist:"
                for version in versions:
                    is_active = ""
                    if version.exists_alias(name=index._name):
                        is_active = self.style.SUCCESS("Active")
                    count = f" ({version.search().count()} documents)"
                    details += f"\n    - {version._name} {is_active}{count}"
            elif exists:
                details = f" ({index.search().count()} documents)"
            else:
                details = ""
            result[module].append(f"{checkbox} {index._name}{details}")
        for app, indices in result.items():
            self.stdout.write(self.style.MIGRATE_LABEL(app))
            self.stdout.write("\n".join(indices))

    def _manage_index(self, action, indices, versioned, version, force, verbosity, ignore_error, **options):  # noqa
        """Manage the creation and deletion of indices."""
        action = OpensearchAction(action)
        known = registry.get_indices()

        # Filter indices
        if indices:
            # Ensure every given indices exists
            known_name = [i._name for i in known]  # noqa
            unknown = set(indices) - set(known_name)
            if unknown:
                self.stderr.write(f"Unknown indices '{list(unknown)}', choices are: '{known_name}'")
                exit(1)

            # Only keep given indices
            indices = list(filter(lambda i: i._name in indices, known))  # noqa
        else:
            indices = known

        # Display expected action
        if verbosity or not force:
            self.stdout.write(f"The following indices will be {action.past}:")
            for index in indices:
                self.stdout.write(f"\t- {index._name}.")  # noqa
            self.stdout.write("")

        # Ask for confirmation to continue
        if not force:  # pragma: no cover
            while True:
                p = input("Continue ? [y]es [n]o : ")
                if p.lower() in ["yes", "y"]:
                    self.stdout.write("")
                    break
                elif p.lower() in ["no", "n"]:
                    exit(1)

        pp = action.present_participle.title()
        for index in indices:
            if verbosity:
                self.stdout.write(
                    f"{pp} index '{index._name}'...\r",
                    ending="",
                )  # noqa
                self.stdout.flush()
            try:
                if action == OpensearchAction.CREATE:
                    if versioned:
                        index = index.create_new_version()
                    else:
                        index.create()
                elif action == OpensearchAction.DELETE:
                    if version:
                        index = index.get_version(version)
                    index.delete()
                elif action == OpensearchAction.REBUILD:
                    try:
                        index.delete()
                    except opensearchpy.exceptions.NotFoundError:
                        pass
                    index.create()
                elif action == OpensearchAction.ACTIVATE:
                    index.activate_version(version)
            except opensearchpy.exceptions.NotFoundError:
                if verbosity or not ignore_error:
                    self.stderr.write(f"{pp} index '{index._name}'...{self.style.ERROR('Error (not found)')}")  # noqa
                if not ignore_error:
                    self.stderr.write("exiting...")
                    exit(1)
            except opensearchpy.exceptions.RequestError:
                if verbosity or not ignore_error:
                    self.stderr.write(
                        f"{pp} index '{index._name}'... {self.style.ERROR('Error (already exists)')}"
                    )  # noqa
                if not ignore_error:
                    self.stderr.write("exiting...")
                    exit(1)
            else:
                if verbosity:
                    self.stdout.write(f"{pp} index '{index._name}'... {self.style.SUCCESS('OK')}")  # noqa

    def _manage_document(
        self,
        action,
        indices,
        force,
        filters,
        excludes,
        verbosity,
        parallel,
        count,
        refresh,
        missing,
        activate,
        **options,
    ):  # noqa
        """Manage the creation and deletion of indices."""
        action = OpensearchAction(action)
        known = registry.get_indices()
        filter_ = functools.reduce(operator.and_, (Q(**{k: v}) for k, v in filters)) if filters else None
        exclude = functools.reduce(operator.and_, (Q(**{k: v}) for k, v in excludes)) if excludes else None

        # Filter indices
        if indices:
            # Ensure every given indices exists
            known_name = [i._name for i in known]  # noqa
            unknown = set(indices) - set(known_name)
            if unknown:
                self.stderr.write(f"Unknown indices '{list(unknown)}', choices are: '{known_name}'")
                exit(1)

            # Only keep given indices
            indices = list(filter(lambda i: i._name in indices, known))  # noqa
        else:
            indices = known

        # Ensure every indices needed are created
        not_created = [i._name for i in indices if not i.exists()]  # noqa
        if not_created:
            self.stderr.write(f"The following indices are not created : {not_created}")
            self.stderr.write("Use 'python3 manage.py opensearch list' to list indices' state.")
            exit(1)

        # Check field, preparing to display expected actions
        s = f"The following documents will be {action.past}:"
        kwargs_list = []
        for index in indices:
            # Handle --missing
            exclude_ = exclude
            if missing and action == OpensearchAction.INDEX:
                q = Q(pk__in=[h.meta.id for h in index.search().extra(stored_fields=[]).scan()])
                exclude_ = exclude_ & q if exclude_ is not None else q

            document = index._doc_types[0]()  # noqa
            try:
                kwargs_list.append({"filter_": filter_, "exclude": exclude_, "count": count})
                qs = document.get_queryset(filter_=filter_, exclude=exclude_, count=count).count()
            except FieldError as e:
                model = index._doc_types[0].django.model.__name__  # noqa
                self.stderr.write(f"Error while filtering on '{model}' (from index '{index._name}'):\n{e}'")  # noqa
                exit(1)
            else:
                s += f"\n\t- {qs} {document.django.model.__name__}."

        # Display expected actions
        if verbosity or not force:
            self.stdout.write(s + "\n\n")

        # Ask for confirmation to continue
        if not force:  # pragma: no cover
            while True:
                p = input("Continue ? [y]es [n]o : ")
                if p.lower() in ["yes", "y"]:
                    self.stdout.write("\n")
                    break
                elif p.lower() in ["no", "n"]:
                    exit(1)

        result = "\n"
        for index, kwargs in zip(indices, kwargs_list):
            document = index._doc_types[0]()  # noqa
            if action == OpensearchAction.MIGRATE:
                success, errors = document.migrate(activate=activate)
            else:
                qs = document.get_indexing_queryset(stdout=self.stdout._out, verbose=verbosity, action=action, **kwargs)
                success, errors = document.update(
                    qs, parallel=parallel, refresh=refresh, action=action, raise_on_error=False
                )

            success_str = self.style.SUCCESS(success) if success else success
            errors_str = self.style.ERROR(len(errors)) if errors else len(errors)
            model = document.django.model.__name__

            if verbosity == 1:
                result += f"{success_str} {model} successfully {action.past}, {errors_str} errors:\n"
                reasons = defaultdict(int)
                for e in errors:  # Count occurrence of each error
                    error = e.get(action, {"result": "unknown error"}).get("result", "unknown error")
                    reasons[error] += 1
                for reasons, total in reasons.items():
                    result += f"    - {reasons} : {total}\n"

            if verbosity > 1:
                result += f"{success_str} {model} successfully {action}d, {errors_str} errors:\n {errors}\n"

        if verbosity:
            self.stdout.write(result + "\n")

    def add_arguments(self, parser):
        """Add arguments to parser."""
        parser.formatter_class = argparse.RawTextHelpFormatter
        subparsers = parser.add_subparsers()

        # 'list' subcommand
        subparser = subparsers.add_parser(
            "list",
            help="Show all available indices (and their state) for the current project.",
            description="Show all available indices (and their state) for the current project.",
        )
        subparser.set_defaults(func=self.__list_index)

        # 'manage' subcommand
        subparser = subparsers.add_parser(
            "index",
            help="Manage the creation an deletion of indices.",
            description="Manage the creation an deletion of indices.",
        )
        subparser.set_defaults(func=self._manage_index)
        subparser.add_argument(
            "action",
            type=str,
            help="Whether you want to create, delete or rebuild the indices. Or activate an index version.",
            choices=[
                OpensearchAction.CREATE.value,
                OpensearchAction.DELETE.value,
                OpensearchAction.REBUILD.value,
                OpensearchAction.ACTIVATE.value,
            ],
        )
        subparser.add_argument("--force", action="store_true", default=False, help="Do not ask for confirmation.")
        subparser.add_argument("--ignore-error", action="store_true", default=False, help="Do not stop on error.")
        subparser.add_argument(
            "--versioned", action="store_true", default=False, help="Set a timestamped name to the actual index"
        )
        subparser.add_argument("--version", help="Only for activate action: version name to activate")
        subparser.add_argument(
            "indices",
            type=str,
            nargs="*",
            metavar="INDEX",
            help="Only manage the given indices.",
        )

        # 'document' subcommand
        subparser = subparsers.add_parser(
            "document",
            help="Manage the indexation and creation of documents.",
            description="Manage the indexation and creation of documents.",
            formatter_class=argparse.RawTextHelpFormatter,
        )
        subparser.set_defaults(func=self._manage_document)
        subparser.add_argument(
            "action",
            type=str,
            help="Whether you want to create, delete, rebuild or migrate the indices.",
            choices=[
                OpensearchAction.INDEX.value,
                OpensearchAction.DELETE.value,
                OpensearchAction.UPDATE.value,
                OpensearchAction.MIGRATE.value,
            ],
        )
        subparser.add_argument(
            "-f",
            "--filters",
            type=self.db_filter(parser),
            nargs="*",
            help=(
                "Filter object in the queryset. Argument must be formatted as '[lookup]=[value]', e.g. "
                "'document_date__gte=2020-05-21.\n"
                "The accepted value type are:\n"
                "  - 'None' ('[lookup]=')\n"
                "  - 'float' ('[lookup]=1.12')\n"
                "  - 'int' ('[lookup]=23')\n"
                "  - 'datetime.date' ('[lookup]=2020-10-08')\n"
                "  - 'list' ('[lookup]=1,2,3,4') Value between comma ',' can be of any other accepted value type\n"
                "  - 'str' ('[lookup]=week') Value that didn't match any type above will be interpreted as a str\n"
                "The list of lookup function can be found here: "
                "https://docs.djangoproject.com/en/dev/ref/models/querysets/#field-lookups"
            ),
        )
        subparser.add_argument(
            "-e",
            "--excludes",
            type=self.db_filter(parser),
            nargs="*",
            help=(
                "Exclude objects from the queryset. Argument must be formatted as '[lookup]=[value]', see '--filters' "
                "for more information"
            ),
        )
        subparser.add_argument("--force", action="store_true", default=False, help="Do not ask for confirmation.")
        subparser.add_argument(
            "-i", "--indices", type=str, nargs="*", help="Only update documents on the given indices."
        )
        subparser.add_argument(
            "-c", "--count", type=int, default=None, help="Update at most COUNT objects (0 to index everything)."
        )
        subparser.add_argument(
            "-p",
            "--parallel",
            action="store_true",
            default=False,
            help="Parallelize the communication with Opensearch.",
        )
        subparser.add_argument(
            "-r",
            "--refresh",
            action="store_true",
            default=False,
            help="Make operations performed on the indices immediatly available for search.",
        )
        subparser.add_argument(
            "-m",
            "--missing",
            action="store_true",
            default=False,
            help="When used with 'index' action, only index documents not indexed yet.",
        )
        subparser.add_argument(
            "-a",
            "--activate",
            action="store_true",
            default=False,
            help="When used with 'migrate' action, activate new Index version immediately.",
        )

        self.usage = parser.format_usage()

    def handle(self, *args, **options):
        """Run the command according to `options`."""
        if "func" not in options:  # pragma: no cover
            self.stderr.write(self.usage)
            self.stderr.write(f"manage.py opensearch: error: no subcommand provided.")
            exit(1)

        options["func"](**options)
